executor/src/witgen/jit/quadratic_symbolic_expression.rs

```diff
@@ -0,0 +1,885 @@
+    {
+        fn get(&self, var: &String) -> RangeConstraint<GoldilocksField> {
+            self.get(var).cloned().unwrap_or_default()
+        }
+    }
+
+    #[test]
+    fn unsolvable() {
+        let r = Qse::from(GoldilocksField::from(10)).solve(&());
+        assert!(r.is_err());
+    }
+
+    #[test]
+    fn unsolvable_with_vars() {
+        let x = &Qse::from_known_symbol("X".to_string(), Default::default());
+        let y = &Qse::from_known_symbol("Y".to_string(), Default::default());
+        let constr = x + y - GoldilocksField::from(10).into();
+        // We cannot solve it, but we can also not learn anything new from it.
+        let result = constr.solve(&()).unwrap();
+        assert!(result.complete && result.effects.is_empty());
+        // But if we know the values, we can be sure there is a conflict.
+        assert!(Qse::from(GoldilocksField::from(10)).solve(&()).is_err());
+    }
+
+    #[test]
+    fn solvable_without_vars() {
+        let constr = Qse::from(GoldilocksField::from(0));
+        let result = constr.solve(&()).unwrap();
+        assert!(result.complete && result.effects.is_empty());
+    }
+
+    #[test]
+    fn solve_simple_eq() {
+        let y = Qse::from_known_symbol("y".to_string(), Default::default());
+        let x = Qse::from_unknown_variable("X".to_string());
+        // 2 * X + 7 * y - 10 = 0
+        let two = Qse::from(GoldilocksField::from(2));
+        let seven = Qse::from(GoldilocksField::from(7));
+        let ten = Qse::from(GoldilocksField::from(10));
+        let constr = two * x + seven * y - ten;
+        let result = constr.solve(&()).unwrap();
+        assert!(result.complete);
+        assert_eq!(result.effects.len(), 1);
+        let Effect::Assignment(var, expr) = &result.effects[0] else {
+            panic!("Expected assignment");
+        };
+        assert_eq!(var.to_string(), "X");
+        assert_eq!(expr.to_string(), "(((7 * y) + -10) / -2)");
+    }
+
+    #[test]
+    fn solve_div_by_range_constrained_var() {
+        let y = Qse::from_known_symbol("y".to_string(), Default::default());
+        let z = Qse::from_known_symbol("z".to_string(), Default::default());
+        let x = Qse::from_unknown_variable("X".to_string());
+        // z * X + 7 * y - 10 = 0
+        let seven = Qse::from(GoldilocksField::from(7));
+        let ten = Qse::from(GoldilocksField::from(10));
+        let mut constr = z * x + seven * y - ten.clone();
+        // If we do not range-constrain z, we cannot solve since we don't know if it might be zero.
+        let result = constr.solve(&()).unwrap();
+        assert!(!result.complete && result.effects.is_empty());
+        let z_rc = RangeConstraint::from_range(1.into(), 2.into());
+        let range_constraints: HashMap<String, RangeConstraint<GoldilocksField>> =
+            HashMap::from([("z".to_string(), z_rc.clone())]);
+        // Just solving without applying the update to the known symbolic expressions
+        // does not help either.
```

Comment:
Why not, actually? What do we pass the range constraints for then?

Reply:
I'll extend the comment: The argument to solve is only for unknown variables. I know it's a bit weird, but range constraints are stored with symbolic expressions because they are computed in the complex expressions and computing them always from scratch would be a bit expensive. So storing range constraints for unknown variables in the QSE is not needed, so I didn't do it.